### PR TITLE
847 fix event bb get delete for admin app

### PR DESCRIPTION
**Pull request by ywkim312**
## Description
Added a method to check if the users belongs to all_group_event by parsing the token. If the user is in the group, the user can get or delete the group event.

**Fixes #(add issue number here and remove parentheses)**

[comment]: # (This project only accepts pull requests related to open issues.)
[comment]: # (If suggesting a new feature or change, please discuss it in an issue first.)
[comment]: # (If fixing a bug, there should be an issue describing it with steps to reproduce.)

## Review Time Estimate
_Please give your idea of how soon this pull request needs to be reviewed by selecting one of the options below. This can be based on the criticality of the issue at hand and/or other relevant factors._

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Immediately
- [x] Within a week
- [ ] When possible

## Type of changes
_Please select a relevant option:_

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Other (any another change that does not fall in one of the above categories.)

## Checklist:
_Please select all applicable options:_

[comment]: # (To select your options, please put an 'x' in the all boxes that apply.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [x] I have signed the Rokwire Contributor License Agreement ([CLA](https://rokwire.org/rokwire_cla)). (_Any contributor who is not an employee of the University of Illinois whose official duties include contributing to the Rokwire software, or who is not paid by the Rokwire project, needs to sign the CLA before their contribution can be accepted._)
- [x] I have updated the [CHANGELOG](../CHANGELOG.md).
- [x] I have read the [Contributor Guidelines](../CONTRIBUTING.md).
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] My change requires updating the documentation.
- [ ] I have made necessary changes to the documentation.
- [ ] I have added tests related to my changes.
- [ ] My changes generate no new warnings.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

---

[comment]: # (Template credit: This pull request template is based on Embedded Artistry {https://github.com/embeddedartistry/templates/blob/master/.github/PULL_REQUEST_TEMPLATE.md}, Clowder {https://github.com/clowder-framework/clowder/blob/develop/.github/PULL_REQUEST_TEMPLATE.md}, and TalAter {https://github.com/TalAter/open-source-templates} templates.)

### CODE REVIEW

File: eventservice/api/utils/group_auth.py [hunk 1]
```diff
@@ -80,4 +82,24 @@ def check_permission_access_event(event, include_private_events, group_ids):
...
-    return True
\ No newline at end of file
+    return True
+
+
+def check_all_group_event_admin():
+    is_all_group_event = False
+    if 'user_token' in g and not g.user_token_data.get('anonymous'):
+        if g.user_token_data.get('permissions'):
+            if g.user_token_data.get('permissions').lower() == ALL_GROUP_EVENTS:
```
Comment (bingzhang) [inline eventservice/api/utils/group_auth.py:92]: @shurwit may we know `g.user_token_data.get('permissions')` will return a list or a comma separated string?

Comment (bingzhang) [inline eventservice/api/utils/group_auth.py:92]: suggested `g.user_token_data.get('permissions').lower().find(ALL_GROUP_EVENTS) != -1:`

File: eventservice/api/utils/group_auth.py [hunk 2]
```diff
@@ -80,4 +82,24 @@ def check_permission_access_event(event, include_private_events, group_ids):
...
+                is_all_group_event = True
+
+    return is_all_group_event
+
+
+def check_all_event_admin():
+    is_all_event = False
+    if 'user_token' in g and not g.user_token_data.get('anonymous'):
+        if g.user_token_data.get('permissions'):
+            if g.user_token_data.get('permissions').lower() == ALL_EVENTS:
```
Comment (bingzhang) [inline eventservice/api/utils/group_auth.py:102]:
```suggestion
           g.user_token_data.get('permissions').lower().find(ALL_GROUP_EVENTS) != -1:
```

**Review by bingzhang — CHANGES_REQUESTED** [review]

File: eventservice/api/utils/group_auth.py (hunk 1 above)
Comment (bingzhang) [inline eventservice/api/utils/group_auth.py:92]:
```suggestion
           g.user_token_data.get('permissions').lower().find(ALL_GROUP_EVENTS) != -1:
```

Comment (sandeep-ps) [reply eventservice/api/utils/group_auth.py:92]: Stephen replied in Slack that this is comma-separated string. Please update the code accordingly. Thanks.

Comment (ywkim312) [reply eventservice/api/utils/group_auth.py]: updated

File: eventservice/api/utils/group_auth.py (hunk 2 above)
Comment (ywkim312) [reply eventservice/api/utils/group_auth.py]: updated

Comment (sandeep-ps) [reply eventservice/api/utils/group_auth.py]: Thanks.

**Comment by sandeep-ps** [comment]
Hi, @ywkim312, Could you please resolve the merge conflicts? Thanks.

**Comment by ywkim312** [comment]
> Hi, @ywkim312, Could you please resolve the merge conflicts? Thanks.

fixed

File: CHANGELOG.md [hunk 3]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [1.12.1] - 2021-12-08
 ### Changed
 - Contributions Catalog login callback endpoint to /catalog/auth/callback. [#803](https://github.com/rokwire/rokwire-building-blocks-api/issues/803) 
+- All group event user or all event user can GET and DELETE group event. [#847](https://github.com/rokwire/rokwire-building-blocks-api/issues/847)
```
[the remark this answers is not in the document]
Comment (ywkim312) [reply CHANGELOG.md]: updated sorry about that

File: eventservice/api/utils/group_auth.py [hunk 4]
```diff
@@ -4,6 +4,8 @@
 from flask import g
 
 logger = logging.getLogger(__name__)
+ALL_GROUP_EVENTS = 'all_group-events'
+ALL_EVENTS = 'all-events'
```
Comment (sandeep-ps) [inline eventservice/api/utils/group_auth.py:8]: @ywkim312, Currently, there is only `all_group-events` permission as this feature is limited to group events. I think you could remove this second permission string and the related case below. What do you think?

**Comment by sandeep-ps** [comment]
@bingzhang, could you please take another look at this PR? It looks like you had requested some changes and your approval will be needed before merging this PR.

In my current tests, with `all_group-events` permission, I'm able to delete any group event.

File: eventservice/api/utils/group_auth.py (hunk 4 above)
Comment (ywkim312) [reply eventservice/api/utils/group_auth.py:8]: I made it for the future use but I can delete these

Comment (ywkim312) [reply eventservice/api/utils/group_auth.py]: removed, however, I had couple methods related to this one in events.py file. Should I remove those as well? Or maybe we can keep those methods and just make this line (variable) reactive when the variables actually comes alive so those method can be used?

File: eventservice/api/controllers/events.py [hunk 5]
```diff
@@ -48,16 +49,38 @@
...
     except Exception as ex:
         __logger.exception(msgs.ERR_MSG_GET_GROUP)
         abort(500)
 
     args = request.args
     query = dict()
     try:
-        query = query_params.format_query(args, query, include_private_events, group_ids)
+        # if all group then give the query with all the group event
+        if is_all_event:
```
Comment (ywkim312) [inline eventservice/api/controllers/events.py:77]: @sandeep-ps this is one of the lines that contains the all_event variable

File: eventservice/api/utils/query_params.py [hunk 6]
```diff
@@ -155,15 +156,26 @@ def format_query(args, query, include_private_events=False, group_ids=None):
...
+            query_parts.append({'isGroupPrivate': {'$ne': True}})
         query['$and'] = query_parts
     # UserAuth query
     else:
-        pubic_private_groups_access_parts = {'$or': [
-            {'isGroupPrivate': {'$ne': True}},  # Includes both group public and regular public events
-            {'createdByGroupId': {'$in': group_ids}}]}  # Check for group membership. Also include group private events.
-        query_parts.append(pubic_private_groups_access_parts)
+        if not all_group_event and not all_event:
+            pubic_private_groups_access_parts = {'$or': [
```
Comment (ywkim312) [inline eventservice/api/utils/query_params.py:165]: @sandeep-ps another line using all_event

File: eventservice/api/utils/query_params.py [hunk 7]
```diff
@@ -155,15 +156,26 @@ def format_query(args, query, include_private_events=False, group_ids=None):
 
     # ApiKeyAuth query
     if not include_private_events:
-        query_parts.append({'isGroupPrivate': {'$ne': True}})
+        if not all_group_event and not all_event:
+            query_parts.append({'isGroupPrivate': {'$ne': True}})
```
Comment (ywkim312) [inline eventservice/api/utils/query_params.py:160]: @sandeep-ps another line using all_event

File: eventservice/api/utils/group_auth.py (hunk 4 above)
Comment (sandeep-ps) [reply eventservice/api/utils/group_auth.py]: Hi, @ywkim312, After going through your comments, I think it's better to remove the use of `all-events` permission at all places. If such permission gets added in the future, we can bring these back. What do you think?

Comment (ywkim312) [reply eventservice/api/utils/group_auth.py]: pushed

File: eventservice/api/utils/query_params.py [hunk 8]
```diff
@@ -155,15 +156,26 @@ def format_query(args, query, include_private_events=False, group_ids=None):
...
+                {'isGroupPrivate': {'$ne': True}},  # Includes both group public and regular public events
+                {'createdByGroupId': {'$in': group_ids}}]}  # Check for group membership. Also include group private events.
+            query_parts.append(pubic_private_groups_access_parts)
         query['$and'] = query_parts
+
+    # All_group_events
+    if all_group_event:
+        query_parts.append({'createdByGroupId': {'$exists': True}})
+
+    # All events
```
Comment (sandeep-ps) [inline eventservice/api/utils/query_params.py:175]: This segment can be removed too, right?

File: CHANGELOG.md [hunk 9]
```diff
@@ -8,6 +8,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - Error message text to event building block when it is internal sever error.  [#840](https://github.com/rokwire/rokwire-building-blocks-api/issues/840)
 
+### Changed
+- All group event user or all event user can GET and DELETE group event. [#847](https://github.com/rokwire/rokwire-building-blocks-api/issues/847)
```
[the remark this answers is not in the document]
Comment (ywkim312) [reply CHANGELOG.md:12]: changed

File: eventservice/api/controllers/events.py [hunk 10]
```diff
@@ -48,16 +49,33 @@
 def search():
     group_ids = list()
     include_private_events = False
+
+    is_all_group_event = False
+
+    # check permission if the user is all event or user is all group event
```
Comment (sandeep-ps) [inline eventservice/api/controllers/events.py:55]:
```suggestion
    # check permission if the user has access to all group events.
```

File: eventservice/api/utils/query_params.py (hunk 8 above)
Comment (ywkim312) [reply eventservice/api/utils/query_params.py:175]: removed

File: eventservice/api/utils/query_params.py [hunk 11]
```diff
@@ -17,7 +17,8 @@
 from bson import ObjectId
 
 
-def format_query(args, query, include_private_events=False, group_ids=None):
+def format_query(args, query, include_private_events=False, group_ids=None,
```
[the remark this answers is not in the document]
Comment (ywkim312) [reply eventservice/api/utils/query_params.py]: removed

File: eventservice/api/controllers/events.py (hunk 10 above)
Comment (ywkim312) [reply eventservice/api/controllers/events.py]: changed

File: eventservice/api/controllers/events.py [hunk 12]
```diff
@@ -396,7 +414,20 @@ def patch(event_id):
 
 
 def delete(event_id):
-    auth_middleware.authorize(auth_middleware.ROKWIRE_EVENT_WRITE_GROUPS)
+    can_delete = False
+
+    # check permission if the user is all event or user is all group event
```
Comment (sandeep-ps) [inline eventservice/api/controllers/events.py:419]:
```suggestion
    # check permission if the user has access to all group events.
```

**Review by sandeep-ps — APPROVED** [review]

**Comment by sandeep-ps** [comment]
Thanks for the changes, Yongwook. In my tests, this is working fine. 

Hi, @bingzhang, Could you please complete your review of this PR? Thanks.

**Review by bingzhang — APPROVED** [review]

**Pull request merged**